tests/templates/fetching_ddp_dynbatch_finetuning/finetune.py

```diff
@@ -0,0 +1,349 @@
+
+    # DDP init
+    sb.utils.distributed.ddp_init_group(run_opts)
+
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    # Kept aside for later
+    reporting = {
+        "WER": {
+            "field": "error_rate",
+            "tracker": deepcopy(
+                hparams["error_rate_computer"]()
+            ),  # n_jobs=int(os.environ['WORLD_SIZE']))),
+        },
+        "CER": {
+            "field": "error_rate",
+            "tracker": deepcopy(
+                hparams["cer_computer"]()
+            ),  # n_jobs=int(os.environ['WORLD_SIZE']))),
+        },
+    }
+
+    # Create experiment directory
+    sb.create_experiment_directory(
+        experiment_directory=hparams["output_folder"],
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
+    # Spared: dataset prepare script & its run_on_main - prepare dataio pipeline only
+    datasets = dataio_prepare(hparams)
+
+    # Add: dynamic batching (only in training & validation)
+    dynamic_hparams = hparams["dynamic_batch_sampler"]
+    train_dataloader_opts = hparams["train_dataloader_opts"]
+    valid_dataloader_opts = hparams["valid_dataloader_opts"]
+    train_dataloader_opts = {
+        "batch_sampler": DynamicBatchSampler(
+            datasets["train"],
+            dynamic_hparams["max_batch_len"],
+            num_buckets=dynamic_hparams["num_buckets"],
+            length_func=lambda x: x["length"],
+            shuffle=dynamic_hparams["shuffle_ex"],
+            batch_ordering=dynamic_hparams["batch_ordering"],
+        ),
+        "num_workers": train_dataloader_opts["num_workers"],
+    }
+    valid_dataloader_opts = {
+        "batch_sampler": DynamicBatchSampler(
+            datasets["valid"],
+            dynamic_hparams["max_batch_len"],
+            num_buckets=dynamic_hparams["num_buckets"],
+            length_func=lambda x: x["length"],
+            shuffle=dynamic_hparams["shuffle_ex"],
+            batch_ordering=dynamic_hparams["batch_ordering"],
+        ),
+        "num_workers": valid_dataloader_opts["num_workers"],
+    }
+
+    # We download:
+    # * the tokenizer from URL - https://huggingface.co/speechbrain/asr-crdnn-rnnlm-librispeech/resolve/main/...
+    # * the pretrained LM from HuggingFace - HF: speechbrain/asr-crdnn-rnnlm-librispeech
+    # * the pretrained ASR from the local template checkpoint - local: speechbrain/asr-crdnn-rnnlm-librispeech
+    run_on_main(
+        hparams["pretrainer_tokenizer"].collect_files,
+        kwargs={"fetch_from": FetchFrom.ONLINE},
+    )
+    run_on_main(
+        hparams["pretrainer_LM"].collect_files,
+        kwargs={"fetch_from": FetchFrom.HUGGING_FACE},
+    )
+    hparams["pretrainer_tokenizer"].load_collected(run_opts["device"])
+    hparams["pretrainer_LM"].load_collected(run_opts["device"])
+    # LOCAL fetching takes sources directly from their location
+    hparams["pretrainer_ASR"].collect_files(fetch_from=FetchFrom.LOCAL)
```

Comment:
As seen here, the same code cannot now handle local and external files.

Reply:
We avoid symlink handling through this PR - symlinks are handled in this sub-setting which now does not need to be called, when data is local to begin with. It's a benefit. I.e. when one directly specifies `FetchFrom.LOCAL`.

Yet, I reckon, you see there's a compatability problem with recipes perhaps. However, when `FetchFrom.LOCAL` is not specified, as in existing recipes, there should not be an issue.